File: src/ruby/tormenta20/models/origem.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing an Origem (Character Origin) in Tormenta20.
    #
    # Origins provide starting skills, items, and sometimes unique powers.
    #
    # @example Find origins with unique powers
    #   Origem.with_unique_power
    #
    # @example Get origin skills
    #   soldado = Origem.find("soldado")
    #   soldado.skills  # => ["Fortitude", "Luta"]
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "soldado")
    # @!attribute [rw] name
    #   @return [String] Origin name (e.g., "Soldado")
    # @!attribute [rw] description
    #   @return [String] Origin description
    # @!attribute [rw] items
    #   @return [Array] Starting items
    # @!attribute [rw] benefits
    #   @return [Hash] Benefits with "skills" and "powers" keys
    # @!attribute [rw] unique_power
    #   @return [Hash, nil] Unique power data or nil
    class Origem < Base
      self.table_name = "origens"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true

      # @!method with_unique_power
      #   Filter origins that have a unique power.
      #   @return [ActiveRecord::Relation<Origem>]
      scope :with_unique_power, -> { where.not(unique_power: nil) }

      # Get trained skills from this origin.
      #
      # @return [Array<String>] List of skill names
      def skills
        benefits&.dig("skills") || []
      end

      # Get powers granted by this origin.
      #
      # @return [Array<String>] List of power names
      def powers
        benefits&.dig("powers") || []
      end

      # Convert origin to a hash representation.
      #
      # @return [Hash] Hash with all origin attributes
      def to_h
        {
          id: id,
          name: name,
          description: description,
          items: items,
          benefits: benefits,
          unique_power: unique_power
        }.compact
      end
    end
  end
end

File: src/ts/models/origem.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"
import { getBookReference, type BookReference } from "./concerns/book_referenceable.js"

const TABLE = "origens"

/**
 * An Origem (character origin/background) in Tormenta20.
 *
 * Origins grant starting equipment, skill training, powers, and optionally
 * a unique ability (`uniquePower`).
 *
 * @example
 * ```ts
 * Origem.all()
 * Origem.withUniquePower().all()
 * Origem.find("nobre")
 *
 * const nobre = Origem.find("nobre")!
 * nobre.skills()   // ["Diplomacia", "Nobreza"]
 * nobre.powers()   // ["poder_id"]
 * ```
 */
export class Origem extends BaseModel {
  get description(): string { return this._row.description as string }
  /** Starting equipment descriptors. */
  get items(): unknown[] { return BaseModel.json(this._row.items, []) }
  /** Raw benefits object with `skills` and `powers` arrays. */
  get benefits(): Record<string, unknown> { return BaseModel.json(this._row.benefits, {}) }
  /** ID of the unique power granted by this origin, or `null`. */
  get uniquePower(): string | null { return (this._row.unique_power as string) ?? null }

  /** Returns skill IDs granted by this origin. */
  skills(): string[] { return (this.benefits.skills as string[]) ?? [] }
  /** Returns power IDs granted by this origin. */
  powers(): string[] { return (this.benefits.powers as string[]) ?? [] }
  /** Returns the sourcebook page reference, or `null` if not indexed. */
  bookReference(): BookReference | null { return getBookReference(this.id) }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description,
      items: this.items, benefits: this.benefits, unique_power: this.uniquePower,
    }
  }

  static query(): Query<Origem> { return BaseModel.makeQuery(TABLE, Origem) }
  static all(): Origem[] { return Origem.query().all() }
  static find(id: string): Origem | null { return Origem.query().find(id) }
  static first(): Origem | null { return Origem.query().first() }
  static last(): Origem | null { return Origem.query().last() }
  static count(): number { return Origem.query().count() }
  /** Returns only origins that have a unique ability. */
  static withUniquePower(): Query<Origem> { return Origem.query().where("unique_power IS NOT NULL") }
}
